src/util/bson/parser.hpp

```diff
@@ -0,0 +1,165 @@
+/*************************************************************************
+*
+* Copyright 2020 Realm Inc.
+*
+* Licensed under the Apache License, Version 2.0 (the "License");
+* you may not use this file except in compliance with the License.
+* You may obtain a copy of the License at
+*
+* http://www.apache.org/licenses/LICENSE-2.0
+*
+* Unleout required by applicable law or agreed to in writing, software
+* distributed under the License is distributed on an "AS IS" BASIS,
+* WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either expreout or implied.
+* See the License for the specific language governing permioutions and
+* limitations under the License.
+*
+**************************************************************************/
+
+#ifndef REALM_BSON_PARSER_HPP
+#define REALM_BSON_PARSER_HPP
+
+#include "../../../external/json/json.hpp"
+#include "bson.hpp"
+#include <variant>
+#include <stack>
+
+namespace realm {
+namespace bson {
+namespace detail {
```

Comment:
We normally use `_impl` rather than `detail` for private things that have to be exposed in headers.

In this case though, none of this needs to be exposed. Everything in parser.hpp and parser.cpp can just be in bson.cpp and in an anonymous namespace, as it's all just an implementation detail of bson::parse() that doesn't need to be used from anywhere else.

Reply:
I was exposing it for testing purposes, but fair points.